packages/affiliates/libs/models.js

```diff
@@ -0,0 +1,184 @@
+      return sum + BigInt(val);
+    }, 0n);
+    return percent(data[affiliate], sum, scale).toString();
+  }
+  function percent(val, sum, scale = 10000n) {
+    return (BigInt(val) * scale) / BigInt(sum);
+  }
+
+  // static function
+  function calculateShares(contributions = {}, scale = 1000000n) {
+    const sum = Object.values(contributions).reduce((sum, val) => {
+      return sum + BigInt(val);
+    }, 0n);
+    if (sum == 0n) return {};
+    return Object.entries(contributions).reduce((result, [key, value]) => {
+      result[key] = percent(value, sum, scale).toString();
+      return result;
+    }, {});
+  }
+
+  function snapshot() {
+    return [...addresses.entries()].reduce((result, [key, value]) => {
+      result[key] = calculateShares(value);
+      return result;
+    }, {});
+  }
+
+  return {
+    get,
+    set,
+    create,
+    getOrCreate,
+    attribute,
+    calculateShare,
+    snapshot
+  };
+}
+
+function Balances() {
+  const balances = new Map();
+  function create(addr) {
+    assert(!balances.has(addr), "Already has addr");
+    return set(addr, "0");
+  }
+  function get(addr) {
+    assert(balances.has(addr), "addr does not exist");
+    return balances.get(addr);
+  }
+  function getOrCreate(addr) {
+    try {
+      return get(addr);
+    } catch (err) {
+      return create(addr);
+    }
+  }
+  function set(addr, balance) {
+    assert(balance >= 0n, "balance must be >= 0: " + balance);
+    balances.set(addr, balance);
+    return balance;
+  }
+  function add(addr, amount) {
+    amount = BigInt(amount);
+    assert(amount >= 0n, "amount must be >= 0: " + amount);
+    const balance = getOrCreate(addr);
+    return set(addr, (BigInt(balance) + BigInt(amount)).toString());
+  }
+  function sub(addr, amount) {
+    amount = BigInt(amount);
+    assert(amount >= 0n, "amount must be >= 0: " + amount);
+    const balance = getOrCreate(addr);
+    return set(addr, (BigInt(balance) - BigInt(amount)).toString());
+  }
+  function snapshot() {
+    return [...balances.entries()].reduce((result, [key, value]) => {
+      result[key] = value;
+      return result;
+    }, {});
+  }
+
+  return {
+    get,
+    set,
+    create,
+    getOrCreate,
+    sub,
+    add,
+    snapshot
+  };
+}
+
+// hold any kind of data which can have a history by block number
+// and also has gaps, so that  not every block number is recorded.
+function History() {
+  const history = [];
+  // Used internally, but will insert a block into cache sorted by timestamp
+  function insert(data) {
+    assert(data.blockNumber, "requires blockNumber");
+    const index = lodash.sortedIndexBy(history, data, "blockNumber");
+    history.splice(index, 0, data);
+    return data;
+  }
+  function lookup(blockNumber) {
+    const index = lodash.sortedIndexBy(history, { blockNumber }, "blockNumber");
+    if (history[index] && history[index].blockNumber === blockNumber) return history[index];
```

Comment:
I think this is reasonable, but IIUC, this method returns the previous block if there is no exact match, right?

Reply:
yes returns exact block or closest previous block